# **Week 05 - Dask Programming Assignment**

Name: Satya Sudha 
Course Number: HDS-5230-07 
Assignment Number: Week 05 Assignment

In [ ]:
import dask.array as da
import pandas as pd
import dask.dataframe as dd
import numpy as np
import time
import requests

## Test

In [ ]:
# make a 3 dimensional matrix of ones with size 1000*1000*1000
start_time = time.time()

images = da.ones((1000,1000,1000))

end_time = time.time()
exe_time= end_time - start_time
exe_time
images

dask.array<ones_like, shape=(1000, 1000, 1000), dtype=float64, chunksize=(255, 255, 255), chunktype=numpy.ndarray>

# 1.

In [ ]:
# URL
url = "http://coronadatascraper.com/timeseries.csv.zip"
zip_filename = "timeseries.csv.zip"

# Download the file
response = requests.get(url)
with open(zip_filename, "wb") as file:
    file.write(response.content)

Download the zip from the link.

# 2.

In [ ]:
# path
csv_path = "timeseries.csv"

dtypes = { "cases": "float64", "deaths": "float64", "recovered": "float64", "county": "object",
    "state": "string", "aggregate": "object",
    "city": "object",
    "population": "float64"
} # To Avoid ValueError
df = dd.read_csv(csv_path, dtype= dtypes, parse_dates=["date"], low_memory = False)

df.head()


locationID slug     name    level  city county state  country    lat  long  \
0    iso1:ad   ad  Andorra  country  <NA>   <NA>  <NA>  Andorra  42.55  1.58   
1    iso1:ad   ad  Andorra  country  <NA>   <NA>  <NA>  Andorra  42.55  1.58   
2    iso1:ad   ad  Andorra  country  <NA>   <NA>  <NA>  Andorra  42.55  1.58   
3    iso1:ad   ad  Andorra  country  <NA>   <NA>  <NA>  Andorra  42.55  1.58   
4    iso1:ad   ad  Andorra  country  <NA>   <NA>  <NA>  Andorra  42.55  1.58   

   ...  deaths recovered active  tested  hospitalized  hospitalized_current  \
0  ...     0.0       0.0    NaN     NaN           NaN                   NaN   
1  ...     0.0       0.0    NaN     NaN           NaN                   NaN   
2  ...     0.0       0.0    NaN     NaN           NaN                   NaN   
3  ...     0.0       0.0    NaN     NaN           NaN                   NaN   
4  ...     0.0       0.0    NaN     NaN           NaN                   NaN   

   discharged  icu  icu_current       date  
0         NaN  NaN          NaN 2020-01-22  
1         NaN  NaN          NaN 2020-01-23  
2         NaN  NaN          NaN 2020-01-24  
3         NaN  NaN          NaN 2020-01-25  
4         NaN  NaN          NaN 2020-01-26  

[5 rows x 24 columns]

In [ ]:
df.info()

<class 'dask.dataframe.core.DataFrame'>
Columns: 24 entries, locationID to date
dtypes: datetime64[ns](1), float64(13), string(10)

In [103]:
# https://stackoverflow.com/questions/50355598/how-should-i-get-the-shape-of-a-dask-dataframe
a = df.shape
a[0].compute(),a[1]

(1089472, 24)

Unzipping and loading using dask is done by giving data types and to prevent memory issues when reading we set low memory to False

# 3.1

In [ ]:
# Filter by'country'(USA)
df_states = df[(df["country"] == "United States") & (df["state"].notnull())]
df_states.head()

locationID       slug                   name  level  city  \
240069  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240070  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240071  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240072  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240073  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   

       county   state        country    lat    long  ...  deaths recovered  \
240069   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240070   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240071   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240072   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240073   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   

       active  tested  hospitalized  hospitalized_current  discharged  icu  \
240069    NaN     NaN           NaN                   NaN         NaN  NaN   
240070    NaN     NaN           NaN                   NaN         NaN  NaN   
240071    NaN     NaN           NaN                   NaN         NaN  NaN   
240072    NaN     NaN           NaN                   NaN         NaN  NaN   
240073    NaN     NaN           NaN                   NaN         NaN  NaN   

        icu_current       date  
240069          NaN 2020-01-23  
240070          NaN 2020-01-24  
240071          NaN 2020-01-25  
240072          NaN 2020-01-26  
240073          NaN 2020-01-27  

[5 rows x 24 columns]

Filtering the rows of the Dask DataFrame (df) to include only those rows where the state column is not empty and the nation column has the value "United States"

# 3.2

In [ ]:
df_2 = df_states[(df_states["date"] >= "2020-01-01") & (df_states["date"] <= "2021-02-28")]
df_2.head()

locationID       slug                   name  level  city  \
240069  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240070  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240071  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240072  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   
240073  iso1:us#iso2:us-ak  alaska-us  Alaska, United States  state  <NA>   

       county   state        country    lat    long  ...  deaths recovered  \
240069   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240070   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240071   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240072   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   
240073   <NA>  Alaska  United States  64.24 -152.23  ...     NaN       NaN   

       active  tested  hospitalized  hospitalized_current  discharged  icu  \
240069    NaN     NaN           NaN                   NaN         NaN  NaN   
240070    NaN     NaN           NaN                   NaN         NaN  NaN   
240071    NaN     NaN           NaN                   NaN         NaN  NaN   
240072    NaN     NaN           NaN                   NaN         NaN  NaN   
240073    NaN     NaN           NaN                   NaN         NaN  NaN   

        icu_current       date  
240069          NaN 2020-01-23  
240070          NaN 2020-01-24  
240071          NaN 2020-01-25  
240072          NaN 2020-01-26  
240073          NaN 2020-01-27  

[5 rows x 24 columns]

Further filters the dataset to include only data within the time period January 1, 2020, to February 28, 2021.

In [ ]:
morrs = df_2.groupby("state").agg(
    {"deaths": "sum", "population": "mean"}
).compute()
morrs["per_capita_mortality"] = (
    morrs["deaths"] / morrs["population"]
)

# Rank states
morrs['rank'] = morrs['per_capita_mortality'].rank(ascending=False)

print(morrs.sort_values(by='rank'))

                                 deaths    population  per_capita_mortality  \
state                                                                         
New York                      9781636.0  6.175734e+05             15.838824   
Texas                         2077857.0  2.274187e+05              9.136704   
Michigan                      2107456.0  2.377823e+05              8.862964   
Louisiana                     1217102.0  1.430398e+05              8.508834   
Georgia                       1230519.0  1.550214e+05              7.937739   
Illinois                      2238987.0  2.879959e+05              7.774370   
Mississippi                    504736.0  7.171443e+04              7.038137   
New Jersey                    4609656.0  8.074718e+05              5.708752   
Pennsylvania                  2145790.0  3.765291e+05              5.698869   
Virginia                       653717.0  1.273958e+05              5.131386   
Ohio                          1029731.0  2.626764e+0

Determines the average population, adds up all of the deaths, and then groups the data by state to determine the per-capita mortality for each state. The ratio of all deaths to the average population for each state is then used to calculate the per-capita mortality. Per capita mortality is then used to rank the states, with the greatest mortality ranking first.

# 3.3

Assumption 1: The likelihood of detecting cases and deaths is consistent over the course of the outbreak.

Assumption 2: All detected cases have resolved (that is, reported cases have either recovered or died).

In [ ]:
# Extract month from the date
df_3 = df[(df["date"] >= "2020-01-01") & (df["date"] <= "2021-02-28")]
df_3 = df_3[(df_3["country"] == "United States") & (df_3["state"].notnull())]
df_3["month"] = df_3["date"].dt.to_period("M")

# Calculate the total deaths and cases per state and month
mcfr = df_3.groupby(["state", "month"]).agg({
    "cases": "sum",
    "deaths": "sum"
}).compute()

# Calculate the CFR
mcfr["CFR"] = (mcfr["deaths"] / mcfr["cases"]) * 100

all_months = pd.period_range("2020-01", "2021-02", freq="M")

# Reindex
cfr_matrix = mcfr["CFR"].unstack().reindex(columns=all_months, fill_value=0)

# Display the CFR matrix (50 states x 14 months)
print(cfr_matrix)

                              2020-01    2020-02   2020-03    2020-04  \
state                                                                   
Alaska                            NaN        NaN  1.340996   2.439908   
Alabama                           NaN        NaN  0.670344   3.222166   
Arkansas                          NaN        NaN  0.931025   1.920520   
American Samoa                    NaN        NaN       NaN        NaN   
Arizona                           0.0   0.000000  1.444484   3.728072   
California                        0.0   5.683564  2.173349   3.565176   
Colorado                          NaN        NaN  1.796194   4.443955   
Connecticut                       NaN        NaN  1.933488   6.451799   
Washington, D.C.                  NaN        NaN  1.482291   3.781444   
Delaware                          NaN        NaN  2.335570   3.279130   
Florida                           NaN        NaN  1.403046   2.911739   
Georgia                           NaN        NaN  3

Same process as 3.2 but this time we are calculating Case fatality ratio (CFR) which is an proportion of individuals diagnosed with a disease who die from that disease and is therefore a measure of severity among detected cases.

![WHO](https://www.who.int/images/default-source/health-topics/coronavirus/cfr.png?sfvrsn=c5031cde_1)

Finally we are pivoting data to get the needed matrix to further explore the dataset.





# 3.4

In [ ]:
# Calculate month-to-month changes in CFR for each state
cfr_changes = cfr_matrix.diff(axis=1)  # Calculate the difference across months

# Sum the changes to see how much the CFR increased or decreased over time for each state
total_changes = cfr_changes.abs().sum(axis=1)

ranked_states = total_changes.rank(ascending=False)

# Display the ranking of states based on CFR changes
print(ranked_states.sort_values())

state
Northern Mariana Islands         1.0
Washington                       2.0
Michigan                         3.0
Connecticut                      4.0
United States Virgin Islands     5.0
Massachusetts                    6.0
California                       7.0
New Jersey                       8.0
Pennsylvania                     9.0
New York                        10.0
New Hampshire                   11.0
Arizona                         12.0
Ohio                            13.0
Missouri                        14.0
Louisiana                       15.0
Illinois                        16.0
Rhode Island                    17.0
Indiana                         18.0
Washington, D.C.                19.0
Minnesota                       20.0
Maryland                        21.0
West Virginia                   22.0
Colorado                        23.0
Florida                         24.0
Wisconsin                       25.0
New Mexico                      26.0
Mississippi                     

We obtain monthly variations in each state's Case Fatality Rate (CFR), add up these variations by adding up their absolute values, and then rank the states according to the overall size of the variations. The findings are shown to indicate which states saw the biggest changes in CFR over time.

In [ ]:
total_changes = cfr_changes.sum(axis=1)
print(total_changes.sort_values(ascending=False))

state
United States Virgin Islands    2.220446e-16
Wyoming                         0.000000e+00
Washington                      0.000000e+00
Oregon                          0.000000e+00
Illinois                        0.000000e+00
Texas                           0.000000e+00
Massachusetts                   0.000000e+00
Utah                            0.000000e+00
Wisconsin                       0.000000e+00
California                      0.000000e+00
Arizona                         0.000000e+00
American Samoa                  0.000000e+00
Northern Mariana Islands        0.000000e+00
Nebraska                        0.000000e+00
Hawaii                         -1.511716e-01
West Virginia                  -3.119152e-01
Rhode Island                   -3.219697e-01
North Carolina                 -3.587541e-01
Tennessee                      -4.839911e-01
Maine                          -6.215635e-01
Alabama                        -6.703437e-01
Minnesota                      -7.350004e-01
New 

# 4.

After all the tasks performed I think parallelization or distributed computing makes sense because dataset has 1089472 rows which is large so. Most of the time we did aggregations (like summing deaths and computing average population) which are ideal candidates for parallel processing and dask could speed up the computation.

In conclusion Large-scale datasets benefit from parallelization or distributed computing features because they span numerous years and states and months of operational duration. This will help to increase the computation speed through the utilization of multiple cores or machines.